### ISSUE TITLE
Smart Map Updates

### ISSUE THREAD

**Issue by dew54** [mapsme/omim#12103]
Hi everyone! I've found this article on the web:
https://medium.com/@Dimitryophoto/maps-me-launches-smart-map-updates-intermediate-stops-and-fast-routes-2a122ef0347

I really want to update my maps just downloading from my map server only the changes, not the entire map.. The article says that MAPS.ME official map already does this.
Is this function present in the open source code?
Thx, 
Davide

**Comment by mpimenov** [comment]
Take a look at this C++ library that has python bindings:
https://github.com/mapsme/omim/tree/dcee7bc238dd1c0439a2923ac4d76f88e722e032/generator/mwm_diff

You will still need to generate both the old and the new versions of the map before a diff between them can be created.

**Comment by dew54** [comment]
Hi @mpimenov , thanks for your reply! 
Once i have the diff file, how do i apply it on a mwm stored on my phone? (eg. iPhone)
And.. i'm a newbie.. How could i run the command for diff generation?
Thank you very much and have a nice day!
Davide

**Comment by mpimenov** [comment]
Running this code on the phone is not the simplest thing to do. Look at the usage of this function in our project to get a grasp on how we do it: https://github.com/mapsme/omim/blob/dcee7bc238dd1c0439a2923ac4d76f88e722e032/generator/mwm_diff/diff.hpp#L34

The short answer is
- if you download the maps via a custom-built version of our app then it should be as easy as setting the correct links and paths
- if you download the maps via a completely different app, you'll have to find out a way to use the mwm-diff library from your code
- if you download the maps manually, you're going to need a separate program to apply the diff. One way is to build the project using the PYBINDINGS option (see here https://github.com/mapsme/omim/blob/dcee7bc238dd1c0439a2923ac4d76f88e722e032/CMakeLists.txt#L71) and then write a small python program

**Comment by dew54** [comment]
Ok so, even if i use a custom built version of your map i still have to generate diffs by myself? And eventually where should i put them? (on server).
Thanks a lot

**Comment by mpimenov** [comment]
Yes, you will need to generate diffs for your custom maps. Put them wherever you want but make sure that your app knows the url. If your directory structure mimicks ours, everything should work out of the box. That is, if you managed to set up the server and your app downloads its maps from your server, it should be easy to adjust everything so that the app downloads diffs.

**Comment by dew54** [comment]
Ok, thank you very much.. one last question.. I really don't know what should i do in order to run the `MakeDiff` function in `diff.hpp`. Is there a way to simply run a command on terminal, pass the mwm and diff paths and have the diff file? I know that in omim-build-release are compiled files, so i thought that just typing a command is enough.
Is it correct? If yes, wich command should i use?
And if not, please, will you so kind and tell me what's needed in order to genersate diffs?
Really, really thank you.
Davide

**Comment by mpimenov** [comment]
For this you either need a separate binary that accepts the flags you want, or a python program.

Example binary from the project:
https://github.com/mapsme/omim/tree/dcee7bc238dd1c0439a2923ac4d76f88e722e032/search/search_quality/features_collector_tool

Example python program from the project:
https://github.com/mapsme/omim/blob/dcee7bc238dd1c0439a2923ac4d76f88e722e032/traffic/pytraffic/example.py

**Comment by dew54** [comment]
Ok thank you very much!
So i wrote this simple main:
`#include "diff.hpp"


using namespace generator::mwm_diff;


int main(int argc, char * argv[])
{

MakeDiff("/Users/davidevidmar/Desktop/Trieste_old.mwm","/Users/davidevidmar/Desktop/Trieste_new.mwm","/Users/davidevidmar/Desktop/");

}`

When i try to compile, g++ generates errors, it seems it can't find files in the includes:
`Air-di-davide:mwm_diff davidevidmar$ g++ mwm_diff_main.cpp diff.cpp -o diff 
In file included from mwm_diff_main.cpp:2:
./diff.hpp:14:6: warning: scoped enumerations are a C++11 extension [-Wc++11-extensions]
enum class DiffApplicationResult
     ^
1 warning generated.
diff.cpp:1:10: fatal error: 'generator/mwm_diff/diff.hpp' file not found
#include "generator/mwm_diff/diff.hpp"`

I've tried to set the include paths in diff.cpp as absolute, but other includes in other files still be relative, and they give me errors.

Please, any suggestion?

**Comment by mpimenov** [comment]
Don't set the paths manually, let CMake do this for you.
Here is a plan:
1. Make sure that you can build and launch a binary from the project.
2. Alter this binary and its CMakeLists.txt until you get the result you want.

**Comment by dew54** [comment]
Ok so, i'm trying to build the file you pointed me out, `features_collection_tools.cpp`
Just typing in his directory on terminal `cmake ./` it says that first, there is no minimum version in CMakeList.txt, so i added this line : `    cmake_minimum_required(VERSION 3.10)`
But then i try again and it says me that:

`  Unknown CMake command "omim_add_executable".


-- Configuring incomplete, errors occurred!
See also "/home/dev/omim/search/search_quality/features_collector_tool/CMakeFiles/CMakeOutput.log".`

And i don't know what's wrong.. 

Maybe i'm missing something?

**Comment by mpimenov** [comment]
You should not invoke CMake in this directory directly. This file is included from the project's main CMakeLists.txt: the parent cmake file provides it with everything it needs (such as omim_add_executable).

Try this:
```
$ cd /home/dev/omim
$ mkdir ../build
$ cd ../build
$ cmake ../omim
$ make -j4 features_collector_tool
```

To make this process more efficient in the future, consider familiarizing yourself with https://cmake.org/documentation/

**Comment by dew54** [comment]
@mpimenov , Thank you man! I did it!
I executed my little c++ program and generated the .diff file! :)

But now i should test it, i have to put it on server.
You said that 
> If your directory structure mimicks ours, everything should work out of the box

Please, could you tell me wich is the correct directory structure?

And, in private.h i guess the url for diffs is in:

`#define DIFF_LIST_URL "" `

Is it right?

Thank you very much! I'm learning new things with your help :)

**Comment by mpimenov** [comment]
Yes, DIFF_LIST_URL is one of the parameters you could use. Read the code that uses it and the tests for it to find out the structure.

**Comment by dew54** [comment]
Ok, i'm lookin into the project, ios and android both, but i can't find anything like an URL or a folder structure.. please could you tell me something more specifical? 

Should the .diff files have a particular name?
And how do i notify to maps.me app that a new version of the map is avaiable?

**Comment by mpimenov** [comment]
https://github.com/mapsme/omim/issues/12147
